chain/chain/src/blocks_delay_tracker.rs

```diff
@@ -0,0 +1,184 @@
+            received_timestamp: timestamp,
+            orphaned_timestamp: None,
+            missing_chunks_timestamp: None,
+            removed_from_orphan_timestamp: None,
+            removed_from_missing_chunks_timestamp: None,
+            height,
+            chunks,
+        });
+    }
+
+    pub fn mark_block_orphaned(&mut self, block_hash: &CryptoHash, timestamp: Instant) {
+        if let Some(block_entry) = self.blocks_in_progress.get_mut(block_hash) {
+            block_entry.orphaned_timestamp = Some(timestamp);
+        } else {
+            debug_assert!(false);
+            warn!(target:"blocks_delay_tracker", "block {:?} was orphaned but was not marked received", block_hash);
+        }
+    }
+
+    pub fn mark_block_unorphaned(&mut self, block_hash: &CryptoHash, timestamp: Instant) {
+        if let Some(block_entry) = self.blocks_in_progress.get_mut(block_hash) {
+            block_entry.removed_from_orphan_timestamp = Some(timestamp);
+        } else {
+            debug_assert!(false);
+            warn!(target:"blocks_delay_tracker", "block {:?} was unorphaned but was not marked received", block_hash);
+        }
+    }
+
+    pub fn mark_block_has_missing_chunks(&mut self, block_hash: &CryptoHash, timestamp: Instant) {
+        if let Some(block_entry) = self.blocks_in_progress.get_mut(block_hash) {
+            block_entry.missing_chunks_timestamp = Some(timestamp);
+        } else {
+            debug_assert!(false);
+            warn!(target:"blocks_delay_tracker", "block {:?} was marked as having missing chunks but was not marked received", block_hash);
+        }
+    }
+
+    pub fn mark_block_completed_missing_chunks(
+        &mut self,
+        block_hash: &CryptoHash,
+        timestamp: Instant,
+    ) {
+        if let Some(block_entry) = self.blocks_in_progress.get_mut(block_hash) {
+            block_entry.removed_from_missing_chunks_timestamp = Some(timestamp);
+        } else {
+            debug_assert!(false);
+            warn!(target:"blocks_delay_tracker", "block {:?} was marked as having no missing chunks but was not marked received", block_hash);
+        }
+    }
+
+    pub fn mark_chunk_received(&mut self, chunk_hash: &ChunkHash, timestamp: Instant) {
+        self.chunks_in_progress
+            .get_mut(&chunk_hash)
+            .map(|chunk_in_progress| chunk_in_progress.chunk_received.get_or_insert(timestamp));
+    }
+
+    pub fn mark_chunk_requested(
+        &mut self,
+        chunk_hash: &ChunkHash,
+        timestamp: Instant,
+        requestor_block_hash: &CryptoHash,
+    ) {
+        self.chunks_in_progress.entry(chunk_hash.clone()).or_insert_with(|| ChunkInProgress {
+            chunk_requested: timestamp,
+            chunk_received: None,
+            block_hash: requestor_block_hash.clone(),
+        });
+    }
+
+    pub fn finish_block_processing(&mut self, block_hash: &CryptoHash, chunks: &[ChunkHash]) {
```

Comment:
(pretty uninformed comment, take with a grain of salt)

It's curious that we update metrics only when we finish block processing. That is, if we never finish a block, it bypasses our metrics. Not sure if that is being done elsewhere, but, eg, logging the len of two hash maps every time we process a block might be interesting

Reply:
Yeah actually I think most other metrics are updated instantly. This is one of the few cases where we gather data and print it at the end of processing blocks.